gr-network/lib/tcp_source_impl.cc

```diff
@@ -0,0 +1,329 @@
+            // and exit loop
+            failed = false;
+            d_host_addr_len = i->ai_addrlen;
+            memcpy(&d_host_addr, i->ai_addr, d_host_addr_len);
+            break;
+        } else {
+            d_socket = socket(
+                host_info->ai_family, host_info->ai_socktype, host_info->ai_protocol);
+            if (d_socket == -1) {
+#ifdef WINDOWS_SOCKETS
+                d_logger->warn("Failed to initialize socket: {}", WSAGetLastError());
+#else
+                d_logger->warn("Failed to initialize socket: {}", strerror(errno));
+#endif
+                failed = true;
+                continue;
+            }
+
+            failed = false;
+            d_host_addr_len = i->ai_addrlen;
+            memcpy(&d_host_addr, i->ai_addr, d_host_addr_len);
+        }
+    }
+
+    if (failed) {
+        d_logger->error("Could not find a usable address, aborting.");
+        throw std::runtime_error(
+            "TCP Source could encountered an error while initializing socket");
+    }
+
+    d_logger->info("Successfully initialized socket.");
+}
+
+tcp_source_impl::~tcp_source_impl()
+{
+
+#ifdef WINDOWS_SOCKETS
+    closesocket(d_socket);
+    WSACleanup();
+#else
+    close(d_socket);
+#endif
+}
+
+
+void tcp_source_impl::establish_connection()
+{
+    if (d_connection_mode == TCP_SOURCE_MODE_CLIENT) {
+        int res = connect(d_socket, (struct sockaddr*)&d_host_addr, d_host_addr_len);
+
+        if (res < 0) {
+#ifdef WINDOWS_SOCKETS
+            d_logger->error("Could not connect to specified server: {}",
+                            WSAGetLastError());
+#else
+            d_logger->error("Could not connect to specified server: {}", strerror(errno));
+#endif
+        } else {
+            d_logger->info("Successfully connected to server.");
+            d_connected = true;
+            d_exchange_socket = d_socket;
+            return;
+        }
+    } else {
+        d_exchange_socket = accept(d_socket, NULL, NULL);
+        if (d_exchange_socket < 0) {
+#ifdef WINDOWS_SOCKETS
+            d_logger->error("Could not accept connection: {}", WSAGetLastError());
+#else
+            d_logger->error("Could not accept connection: {}", strerror(errno));
+#endif
+            return;
+        }
+
+        d_logger->info("Successfully connected to Client.");
+        d_connected = true;
+    }
+
+    return;
+}
+
+int tcp_source_impl::work(int noutput_items,
+                          gr_vector_const_void_star& input_items,
+                          gr_vector_void_star& output_items)
+{
+
+    while (!d_connected) {
+        // if we're not connected, then attempt to connect with something, also
+        // block until we succeed
+        this->establish_connection();
+    }
+
+    int* out = (int*)output_items[0];
+
+    int to_be_received = d_block_size * noutput_items;
+
+    // we have connection, good, then we receive.
+    while (to_be_received > 0) {
```

Comment:
But we should return as soon as we got something, right? Don't want to block until the output buffer is full?

Reply:
I think we do, since we have a scheduler that tells us how many samples we need to produce each call. And since we're one of the sources in the flowgraph I assume its best to follow that to the letter. The only exception for this is if the connection gets terminated, since we can't really magically produce samples out of thin air

Reply:
The scheduler tells you how many you *can* produce, the maximum the output has space for.

That's at least my understanding. Though I can't immediately find any documentation saying so, e.g. in https://wiki.gnuradio.org/index.php/BlocksCodingGuide

Reply:
I quick glance of the UDP one it looks like it doesn't block, but adds all the packets that are there right now, and then returns.

This makes the data available to downstream blocks ASAP, which is what I mean is desirable, so that the rest of the flowgraph isn't so stop-go.

Reply:
So it's not an issue either way, good to know

Reply:
Alright. I'm not in the approve path for this, but I would not approve it. It's assuming that new data comes in within a reasonable time, and I don't find that to be a safe assumption.

I'll reserve the right to say "I told you so" when people report bugs about stuck data.